File: src/CsRopExample/MessageLoggingHandler.cs
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Threading;

namespace CsRopExample
{
    /// <summary>
    /// Logging code 
    /// </summary>
    public class MessageLoggingHandler : MessageProcessingHandler
    {

        protected override HttpRequestMessage ProcessRequest(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            var correlationId = string.Format("{0}{1}", DateTime.Now.Ticks, Thread.CurrentThread.ManagedThreadId);
            var requestInfo = string.Format("{0} {1}", request.Method, request.RequestUri);
            var message = request.Content.ReadAsStringAsync().Result;
            Debug.WriteLine("{0} - Request: {1}\r\n{2}", correlationId, requestInfo, message);
            return request;
        }

        protected override HttpResponseMessage ProcessResponse(HttpResponseMessage response, CancellationToken cancellationToken)
        {
            var correlationId = string.Format("{0}{1}", DateTime.Now.Ticks, Thread.CurrentThread.ManagedThreadId);
            var requestInfo = string.Format("{0} {1}", response.RequestMessage.Method, response.RequestMessage.RequestUri);
            var message = response.Content != null 
                ? response.Content.ReadAsStringAsync().Result 
                : "[no body]";
            Debug.WriteLine("{0} - Response: {1}\r\n{2}", correlationId, requestInfo, message);
            return response;
        }
    }
}

File: src/FsRopExample/MessageLoggingHandler.fs
namespace FsRopExample

open System
open System.Diagnostics
open System.Net.Http
open System.Threading

/// Logging code 
type MessageLoggingHandler() =
    inherit MessageProcessingHandler()

    override this.ProcessRequest(request:HttpRequestMessage , _:CancellationToken) =
        let correlationId = sprintf "%i%i" DateTime.Now.Ticks Thread.CurrentThread.ManagedThreadId
        let requestInfo = sprintf "%O %O" request.Method  request.RequestUri
        let message = request.Content.ReadAsStringAsync().Result
        Debug.WriteLine("[HTTP]Request: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message)
        request

    override this.ProcessResponse(response:HttpResponseMessage, _:CancellationToken) =
        let correlationId = sprintf "%i%i" DateTime.Now.Ticks Thread.CurrentThread.ManagedThreadId
        let requestInfo = sprintf "%O %O" response.RequestMessage.Method  response.RequestMessage.RequestUri
        let message = if response.Content <> null then response.Content.ReadAsStringAsync().Result else "[no body]"
        Debug.WriteLine("[HTTP]Response: {1}\r\n[HTTP]{2}\r\n\r\n", correlationId, requestInfo, message)
        response
